cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java

```diff
@@ -496,28 +498,39 @@ private Location createTempLocationDirectory() throws IOException {
    */
   @SuppressWarnings("unchecked")
   private void beforeSubmit(final Job job) throws TransactionFailureException {
+
+    final TransactionControl txControl = mapReduce instanceof AbstractMapReduce
```

Comment:
AbstractMapReduce implements initialize(context) as final and requires that subclasses implement initialize(), without arguments. For destroy, AbstractMapReduce simply defines a no-op method. So it does not matter whether the program implements AbstractMapReduce.